src/rust/engine/src/scheduler.rs

```diff
@@ -15,6 +16,29 @@ use nodes::{NodeKey, Select};
 use rule_graph;
 use selectors;
 
+pub struct Session {
+  // The set of roots that have been requested within this session.
```

Comment:
Can you add a comment explaining why we're using interior mutability here rather than `&mut self` on `extend`?